### DISCUSSION TITLE
Doubt about webauthn4j

### DISCUSSION THREAD

**Discussion by ovninosa** [webauthn4j/webauthn4j#706] [Q&A] [answered]
Hey @ynojima how are you?

First of all thanks for your work.
I have just a simple question but I cannot figured it out myself.

What are the main difference between webauthn4j vs yubico fido server vs line fido server?
I think there are the same idea and uses cases, but I cannot found detailed info about this.

Thanks!

**Comment by ynojima** [comment, marked as answer]
Good question.
I'd like to hear the thought from the author of yubico fido server and line fido server too :)
I think they have another opinions.

Actually, I haven't used yubico fido server and line fido server, and I'm not sure about the details of them,
but I will answer the question as far as my understanding allows.

## Server or Library

line fido server is a server, not library. It run as a separate server process and main application process will communicate with it via REST API.
The REST API exposes OpenAPI schema, so main application process can be written in non-JVM language.

WebAuthn4J is a Java library, and can be run in the main application process. It can be called from JVM language only, but It exposes less overhead. It is easy to debug.
yubico fido server is a Java library too.

## Dependencies

yubico fido server and line fido server depends on various 3rd-party libraries.
WebAuthn4J depends on just few libraries at runtime:
- SLF4J (for logging)
- Jackson (for JSON/CBOR parsing)
- Apache Kerby(for ASN.1 parsing)

Less is better from the point of update cost, and security risk, I think.

## MDS3 support 

WebAuthn4J and yubico fido server support it. line fido server doesn't.

## Attestation Statement support

WebAuthn4J and line fido server supports all attestation formats.
yubico fido server doesn't support `tpm` and `android-key`.

## Documentation

It seems yubico fido server is well documented.
Due to my laziness, WebAuthn4J is not, but please ask me anything :)

**Reply by ovninosa** [reply]
amazing dude, thanks for all the details !
